pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,33 +318,73 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
 	return potentialNodes
 }
 
+type candidateList struct {
+	idx int32
+	max int32
+	arr []Candidate
+}
+
+func newCandidateList(max int32) *candidateList {
+	return &candidateList{max: max, idx: -1, arr: make([]Candidate, max)}
+}
+
+// add adds a new candidate to the internal array atomically.
+func (cl *candidateList) add(c *candidate) {
+	if idx := atomic.AddInt32(&cl.idx, 1); idx < cl.max {
+		cl.arr[idx] = c
+	} else {
+		atomic.AddInt32(&cl.idx, -1)
+	}
+}
+
+// size returns the number of candidate stored atomically.
+func (cl *candidateList) size() int32 {
+	return atomic.LoadInt32(&cl.idx) + 1
+}
+
+// get returns the internal candidate array. This function is NOT atomic and
+// assumes that the caller has acquired mutual exclusion.
+func (cl *candidateList) get() []Candidate {
+	return cl.arr[:cl.idx+1]
```

Comment:
why do we want a copy? this is getting appended to create a slice anyways at line 387

Reply:
This is just a slice, not a copy. If there are fewer than `cl.idx+1`, then `cl.arr` will have nil padding, which we don't want to append (`arr` has len = cap).

Reply:
Sg